merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/CheckpointSimulationDriver.java

```diff
@@ -0,0 +1,412 @@
+    MutableObject<Duration> lastCheckpointSimTime = new MutableObject<>(Duration.ZERO);
+    return simulationState -> {
+      if (simulationState.nextTime().longerThan(simulationState.currentTime()) && System.nanoTime() - lastCheckpointRealTime.getValue() > (thresholdSeconds * 1000 * 1000 * 1000)) {
+        lastCheckpointRealTime.setValue(System.nanoTime());
+        lastCheckpointSimTime.setValue(simulationState.currentTime());
+        return true;
+      } else {
+        return false;
+      }
+    };
+  }
+
+  public static Function<SimulationState, Boolean> checkpointAtEnd(Function<SimulationState, Boolean> stoppingCondition) {
+    return simulationState -> stoppingCondition.apply(simulationState) || simulationState.nextTime.isEqualTo(MAX_VALUE);
+  }
+
+  public record SimulationState(
+      Duration currentTime,
+      Duration nextTime,
+      SimulationEngine simulationEngine,
+      Map<ActivityDirectiveId,  ActivityDirective> schedule,
+      Map<ActivityDirectiveId, SpanId> activityDirectiveIdSpanIdMap
+  ){}
+
+  public static <Model> SimulationResultsComputerInputs simulateWithCheckpoints(
+      final MissionModel<Model> missionModel,
+      final Map<ActivityDirectiveId, ActivityDirective> schedule,
+      final Instant simulationStartTime,
+      final Duration simulationDuration,
+      final Instant planStartTime,
+      final Duration planDuration,
+      final Consumer<Duration> simulationExtentConsumer,
+      final Supplier<Boolean> simulationCanceled,
+      final CachedSimulationEngine cachedEngine,
+      final Function<SimulationState, Boolean> shouldTakeCheckpoint,
+      final Function<SimulationState, Boolean> stopConditionOnPlan,
+      final CachedEngineStore cachedEngineStore,
+      final SimulationEngineConfiguration configuration,
+      final boolean duplicationIsOk) {
+    final var activityToSpan = new HashMap<ActivityDirectiveId, SpanId>();
+    final var activityTopic = cachedEngine.activityTopic();
+    final var timelines = new ArrayList<TemporalEventSource>();
+    timelines.add(new TemporalEventSource(cachedEngine.timePoints));
+    var engine = !duplicationIsOk ? cachedEngine.simulationEngine : cachedEngine.simulationEngine.duplicate();
+    engine.unscheduleAfter(cachedEngine.endsAt);
+
+    var timeline = new TemporalEventSource();
+    var cells = new LiveCells(timeline, cachedEngine.cells());
+    /* The current real time. */
+    var elapsedTime = Duration.max(ZERO, cachedEngine.endsAt());
+
+    simulationExtentConsumer.accept(elapsedTime);
+
+    // Specify a topic on which tasks can log the activity they're associated with.
+
+    try {
+      final var filteredSchedule = new HashMap<ActivityDirectiveId, ActivityDirective>();
+      for (final var entry : schedule.entrySet()) {
+        if (entry.getValue().startOffset().longerThan(cachedEngine.endsAt())) {
+          filteredSchedule.put(entry.getKey(), entry.getValue());
+        }
```

Comment:
I have implemented StartOffsetReducer.filterOutStartOffsetsBefore(resolvedMap, Duration). You'll see that we use it with cachedEngine.endsAt().plus(MICROSECONDS) because the method implements filters out values < than Duration but we want <= in the case of the cached engine;